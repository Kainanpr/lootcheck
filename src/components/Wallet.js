import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { makeDeposit, makeWithdraw } from '../redux/actions/balance';

export class Wallet extends Component {
  state = {
    inputValue: undefined,
  }

  static propTypes = {
    balance: PropTypes.number,
    makeDeposit: PropTypes.func,
  };

  static defaultProps = {
    balance: undefined,
    makeDeposit: () => {},
  };

  changeInputValue = (event) => {
    this.setState({ inputValue: Number(event.target.value) });
  }

  makeDeposit = () => {
    const { inputValue } = this.state;
    const { makeDeposit } = this.props;

    makeDeposit(inputValue);
  }

  makeWithdraw = () => {
    const { inputValue } = this.state;
    const { makeWithdraw } = this.props;

    makeWithdraw(inputValue);
  }

  render() {
    const { balance } = this.props;
    const { inputValue } = this.state;

    return (
      <div>
        <h3 className="balance">Wallet balance: {balance}</h3>
        <br />
        <input
          className="input-wallet"
          onChange={this.changeInputValue}
          value={inputValue}
        />
        <button
          type="button"
          className="btn-deposit"
          onClick={this.makeDeposit}
        >
          Deposit
        </button>
        <button
          type="button"
          className="btn-withdraw"
          onClick={this.makeWithdraw}
        >
          Withdraw
        </button>
      </div>
    );
  }
}

const mapStateToProps = state => (
  {
    balance: state,
  }
);

const mapDispatchToProps = {
  makeDeposit,
  makeWithdraw,
};

export default connect(mapStateToProps, mapDispatchToProps)(Wallet);
