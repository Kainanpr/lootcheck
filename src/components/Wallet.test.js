import React from 'react';
import { shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { Wallet } from './Wallet';

configure({ adapter: new Adapter() });

describe('Wallet', () => {
  const mockMakeDeposit = jest.fn();
  const props = { balance: 20, makeDeposit: mockMakeDeposit };
  const wrapper = shallow(<Wallet {...props} />);

  it('renders properly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('displays the balance from props', () => {
    expect(wrapper.find('.balance').text()).toEqual('Wallet balance: 20');
  });

  it('creates an input to deposit into or withdraw from the balance', () => {
    expect(wrapper.find('.input-wallet').exists()).toBe(true);
  });

  describe('when the user types into the wallet input', () => {
    const userBalance = '25';

    beforeEach(() => {
      wrapper.find('.input-wallet')
        .simulate('change', { target: { value: userBalance } });
    });

    it('updates the local wallet balance in `state` and converts it to a number', () => {
      expect(wrapper.state().inputValue).toEqual(Number(userBalance));
    });

    describe('and the user wants to make a deposit', () => {
      beforeEach(() => {
        wrapper.find('.btn-deposit').simulate('click');
      });

      it('dispatches the `makeDeposit()` it receives from props with local balance', () => {
        expect(mockMakeDeposit).toHaveBeenCalledWith(Number(userBalance));
      });
    });
  });
});
