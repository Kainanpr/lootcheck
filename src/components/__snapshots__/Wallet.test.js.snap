// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Wallet renders properly 1`] = `
<div>
  <h3
    className="balance"
  >
    Wallet balance: 
    20
  </h3>
  <br />
  <input
    className="input-wallet"
    onChange={[Function]}
  />
  <button
    className="btn-deposit"
    onClick={[Function]}
    type="button"
  >
    Deposit
  </button>
  <button
    className="btn-withdraw"
    onClick={[Function]}
    type="button"
  >
    Withdraw
  </button>
</div>
`;
